File: backend/app/services/packages.py
"""Работа с базой пакетов: поиск, дедупликация, получение/создание версий."""

from __future__ import annotations

from typing import Any

from sqlalchemy import func, or_, select
from sqlalchemy.orm import Session

from app.core.config import get_settings
from app.db.enums import PENDING_VERSION_STATUSES, STATUS_TITLES
from app.db.models import Package, PackageVersion, RequestItem
from app.managers.base import PackageRef
from app.managers.registry import get_plugin

APPROVED_STATUSES = ("approved",)


def find_version(session: Session, ref: PackageRef) -> PackageVersion | None:
    stmt = (
        select(PackageVersion)
        .join(Package)
        .where(
            Package.manager == ref.manager,
            Package.name == ref.name,
            PackageVersion.version == ref.version,
        )
    )
    return session.execute(stmt).scalars().first()


def get_or_create_version(session: Session, ref: PackageRef) -> tuple[PackageVersion, bool]:
    """Возвращает версию пакета и признак «создана только что»."""
    package = session.execute(
        select(Package).where(Package.manager == ref.manager, Package.name == ref.name)
    ).scalars().first()
    if package is None:
        package = Package(manager=ref.manager, name=ref.name, display_name=ref.display_name)
        session.add(package)
        session.flush()

    version = session.execute(
        select(PackageVersion).where(
            PackageVersion.package_id == package.id, PackageVersion.version == ref.version
        )
    ).scalars().first()
    if version is not None:
        return version, False

    version = PackageVersion(
        package_id=package.id, version=ref.version, raw_version=ref.raw_version, status="new"
    )
    session.add(version)
    session.flush()
    return version, True


def install_command(session: Session, version: PackageVersion) -> str:
    s = get_settings()
    plugin = get_plugin(version.package.manager)
    ref = version_ref(version)
    return plugin.install_command(ref, s.artifact_base_url, s.artifact_repo(version.package.manager))


def version_ref(version: PackageVersion) -> PackageRef:
    return PackageRef(
        manager=version.package.manager,
        name=version.package.name,
        display_name=version.package.display_name,
        version=version.version,
        raw_version=version.raw_version,
    )


def search_versions(
    session: Session,
    *,
    query: str | None = None,
    manager: str | None = None,
    version: str | None = None,
    status: str | None = None,
    limit: int = 50,
    offset: int = 0,
) -> tuple[list[PackageVersion], int]:
    """Поиск по базе пакетов: имя, версия, менеджер, статус."""
    stmt = select(PackageVersion).join(Package)
    count_stmt = select(func.count()).select_from(PackageVersion).join(Package)
    conditions = []
    if query:
        pattern = f"%{query.strip().lower()}%"
        conditions.append(
            or_(func.lower(Package.name).like(pattern), func.lower(Package.display_name).like(pattern))
        )
    if manager:
        conditions.append(Package.manager == manager)
    if version:
        conditions.append(PackageVersion.version.like(f"{version.strip()}%"))
    if status:
        conditions.append(PackageVersion.status == status)
    for cond in conditions:
        stmt = stmt.where(cond)
        count_stmt = count_stmt.where(cond)

    total = session.execute(count_stmt).scalar_one()
    stmt = stmt.order_by(Package.name, PackageVersion.id.desc()).limit(limit).offset(offset)
    return list(session.execute(stmt).scalars().all()), int(total)


def check_state(session: Session, version: PackageVersion) -> dict[str, Any]:
    """Что разработчику делать с найденной в базе версией.

    Наличие записи само по себе ничего не значит: строка создаётся в момент
    заведения заявки, до всех проверок. Поэтому отдаём не «найдено/не найдено», а
    состояние по существу: можно ставить, идёт проверка или проверку не прошёл.
    """
    status = version.status
    payload: dict[str, Any] = {
        "status": status,
        "status_title": STATUS_TITLES.get(status, status),
        "package_version_id": version.id,
        "link": f"/api/v1/packages/{version.id}",
        "status_reason": version.status_reason,
    }

    if status == "approved":
        payload["state"] = "approved"
        payload["message"] = "Одобрен — можно ставить из внутреннего репозитория."
        payload["install_command"] = install_command(session, version)
        return payload

    if status in PENDING_VERSION_STATUSES:
        # Заявка уже есть — подсказываем её номер, чтобы не заводили вторую.
        item = session.execute(
            select(RequestItem)
            .where(RequestItem.package_version_id == version.id)
            .order_by(RequestItem.id.desc())
        ).scalars().first()
        payload["state"] = "in_progress"
        payload["request_id"] = item.request_id if item else None
        payload["current_step"] = item.current_step if item else None
        payload["next_action"] = item.next_action if item else None
        where = f" (заявка #{item.request_id})" if item else ""
        payload["message"] = (
            f"Ставить нельзя: пакет ещё не прошёл модерацию — {payload['status_title'].lower()}"
            f"{where}. Повторную заявку заводить не нужно."
        )
        return payload

    payload["state"] = "blocked"
    reason = (version.status_reason or "").strip()
    if reason and reason[-1] not in ".!?":
        reason += "."  # причина приходит из разных мест, точку в конце не гарантируем
    payload["message"] = " ".join(
        part
        for part in (
            f"Ставить нельзя: {payload['status_title'].lower()}.",
            reason,
            "Подберите другую версию или замену.",
        )
        if part
    )
    return payload


def version_summary(session: Session, version: PackageVersion) -> dict[str, Any]:
    package = version.package
    payload: dict[str, Any] = {
        "id": version.id,
        "manager": package.manager,
        "name": package.display_name,
        "normalized_name": package.name,
        "version": version.raw_version,
        "status": version.status,
        "license_spdx": version.license_spdx,
        "license_source": version.license_source,
        "published_at": version.published_at,
        "quarantine_until": version.quarantine_until,
        "approved_at": version.approved_at,
        "max_vuln_score": version.max_vuln_score,
        "status_reason": version.status_reason,
        "vulnerabilities": [
            {
                "id": v.external_id,
                "score": v.score,
                "cvss_vector": v.cvss_vector,
                "severity": v.severity,
                "url": v.url,
                "summary": v.summary,
                "fixed_versions": v.fixed_versions,
                "affected_ranges": v.affected_ranges,
                "index_version_id": v.vuln_index_version_id,
            }
            for v in sorted(version.vulnerabilities, key=lambda v: -v.score)
        ],
        "artifacts": [
            {
                "filename": a.filename,
                "nexus_url": a.nexus_url,
                "sha256": a.sha256,
                "size_bytes": a.size_bytes,
                "s3_key": a.s3_key,
                "s3_deleted_at": a.s3_deleted_at,
                "status": a.status,
            }
            for a in version.artifacts
        ],
    }
    if version.status == "approved":
        payload["install_command"] = install_command(session, version)
    return payload

File: backend-go/internal/api/packages.go
package api

import (
	"net/http"
	"strconv"
	"strings"

	"github.com/go-chi/chi/v5"

	"time"

	"moderation/internal/artifactstore"
	"moderation/internal/config"
	"moderation/internal/decisions"
	"moderation/internal/domain"
	"moderation/internal/registry"
	"moderation/internal/repo"
)

// База пакетов и справочник менеджеров. Порт backend/app/api/v1/packages.go.
//
// Формат ответов совпадает с python-версией дословно: эти маршруты читает тот
// же фронтенд, и он переезда не заметит.

// PackagesHandler — зависимости маршрутов базы пакетов.
type PackagesHandler struct {
	Repo     *repo.Repo
	Registry *registry.Registry
	Cfg      *config.Config
	// Decisions — сервис решений, нужен отзыву пакета. nil — маршрут отзыва
	// не подключается: отдать его неработающим хуже, чем не отдать совсем.
	Decisions *decisions.Service
	// Artifacts — артефактори, откуда снимается отозванный пакет. nil
	// допустим: сервис поднимается и без него, и тогда отзыв меняет только
	// статус, о чём ответ говорит прямо.
	Artifacts artifactstore.Store
	Now       func() time.Time
}

func (h *PackagesHandler) now() time.Time {
	if h.Now != nil {
		return h.Now()
	}
	return time.Now().UTC()
}

// MountPackages подключает справочник менеджеров и базу пакетов. Всё закрыто
// проверкой токена: в базе видно, какие пакеты и версии тянет компания.
// Группа с полными путями, а не Route("/api/v1"): под этим префиксом уже
// смонтированы /auth и /requests, и chi на попытку смонтировать туда же второй
// обработчик падает паникой при сборке роутера.
func MountPackages(r chi.Router, h *PackagesHandler, a *Auth) {
	r.Group(func(sub chi.Router) {
		sub.Use(a.Authenticate)
		sub.Get("/api/v1/managers", h.Managers)
		sub.Get("/api/v1/managers/detect", h.DetectManager)
		sub.Get("/api/v1/packages", h.Search)
		sub.Get("/api/v1/packages/{versionID}", h.Version)
		// «Найти пакет»: отвечает не «есть в базе», а можно ли ставить.
		sub.Post("/api/v1/packages/check", h.Check)
	})
	// Отзыв — только DevSecOps и админу: он отменяет ранее принятое решение и
	// снимает пакет с публикации у всех сразу.
	if h.Decisions != nil {
		r.Group(func(sub chi.Router) {
			sub.Use(a.Authenticate)
			sub.Use(RequireRoles("admin", "devsecops"))
			sub.Post("/api/v1/packages/{versionID}/revoke", h.Revoke)
		})
	}
}

// Managers — GET /api/v1/managers. Порядок совпадает с python-версией
// (pypi, npm, go, nuget): по нему строится выпадающий список.
func (h *PackagesHandler) Managers(w http.ResponseWriter, r *http.Request) {
	plugins := h.Registry.Plugins()
	out := make([]map[string]any, 0, len(plugins))
	for _, p := range plugins {
		out = append(out, map[string]any{
			"code":         p.Code(),
			"title":        p.Title(),
			"entry_format": p.EntryFormat(),
			// У git/files нет файлов зависимостей. Нулевой slice нельзя
			// сериализовать как null: интерфейс обращается с полем как со
			// списком, а контракт /managers обещает именно массив.
			"dependency_files": listOrEmpty(p.DependencyFiles()),
			"osv_ecosystem":    p.OSVEcosystem(),
		})
	}
	writeJSON(w, http.StatusOK, out)
}

// DetectManager — GET /api/v1/managers/detect?filename=go.sum.
// manager = null, если файл ничей: это подсказка интерфейсу, а не ошибка.
func (h *PackagesHandler) DetectManager(w http.ResponseWriter, r *http.Request) {
	filename := r.URL.Query().Get("filename")
	if strings.TrimSpace(filename) == "" {
		writeError(w, r, errValidation("Параметр filename обязателен"))
		return
	}
	var manager any
	if code := h.Registry.DetectByFile(filename); code != "" {
		manager = code
	}
	writeJSON(w, http.StatusOK, map[string]any{"filename": filename, "manager": manager})
}

// Search — GET /api/v1/packages. Поиск по базе пакетов.
func (h *PackagesHandler) Search(w http.ResponseWriter, r *http.Request) {
	q := r.URL.Query()
	limit, err := intParam(q.Get("limit"), 50)
	if err != nil {
		writeError(w, r, errValidation("Параметр limit должен быть числом"))
		return
	}
	if limit > 200 {
		writeError(w, r, errValidation("Параметр limit не больше 200"))
		return
	}
	offset, err := intParam(q.Get("offset"), 0)
	if err != nil || offset < 0 {
		writeError(w, r, errValidation("Параметр offset должен быть неотрицательным числом"))
		return
	}

	rows, total, err := h.Repo.SearchVersions(r.Context(), repo.VersionSearch{
		Query: q.Get("q"), Manager: q.Get("manager"),
		Version: q.Get("version"), Status: q.Get("status"),
		Limit: limit, Offset: offset,
	})
	if err != nil {
		writeError(w, r, errInternal("Не удалось выполнить поиск по базе пакетов").Because(err))
		return
	}

	items := make([]map[string]any, 0, len(rows))
	for i := range rows {
		view, err := h.versionSummary(r, &rows[i])
		if err != nil {
			writeError(w, r, errInternal("Не удалось собрать карточку пакета").Because(err))
			return
		}
		items = append(items, view)
	}
	writeJSON(w, http.StatusOK, map[string]any{
		"total": total, "limit": limit, "offset": offset, "items": items,
	})
}

// Version — GET /api/v1/packages/{versionID}. Карточка версии пакета.
func (h *PackagesHandler) Version(w http.ResponseWriter, r *http.Request) {
	versionID, ok := pathInt64(w, r, "versionID")
	if !ok {
		return
	}
	row, err := h.Repo.GetVersionRow(r.Context(), versionID)
	if err != nil {
		writeError(w, r, errInternal("Не удалось прочитать карточку пакета").Because(err))
		return
	}
	if row == nil {
		writeError(w, r, errNotFound("Версия пакета #"+strconv.FormatInt(versionID, 10)+" не найдена"))
		return
	}
	view, err := h.versionSummary(r, row)
	if err != nil {
		writeError(w, r, errInternal("Не удалось собрать карточку пакета").Because(err))
		return
	}
	writeJSON(w, http.StatusOK, view)
}

// versionSummary — порт services.packages.version_summary.
func (h *PackagesHandler) versionSummary(r *http.Request, row *repo.VersionRow) (map[string]any, error) {
	v := row.Version
	vulns, err := h.Repo.ListVulnerabilities(r.Context(), v.ID)
	if err != nil {
		return nil, err
	}
	artifacts, err := h.Repo.ListArtifacts(r.Context(), v.ID)
	if err != nil {
		return nil, err
	}

	out := map[string]any{
		"id":               v.ID,
		"manager":          row.Manager,
		"name":             row.DisplayName,
		"normalized_name":  row.Name,
		"version":          v.RawVersion,
		"status":           v.Status,
		"license_spdx":     v.LicenseSPDX,
		"license_source":   v.LicenseSource,
		"published_at":     v.PublishedAt,
		"quarantine_until": v.QuarantineUntil,
		"approved_at":      v.ApprovedAt,
		"max_vuln_score":   v.MaxVulnScore,
		"status_reason":    v.StatusReason,
		"vulnerabilities":  vulnerabilityViews(vulns, true),
		"artifacts":        artifactViews(artifacts),
	}
	// Команда установки имеет смысл только у одобренного пакета: у остального
	// она вела бы в репозиторий, где артефакта нет.
	if v.Status == "approved" {
		if cmd := h.installCommand(row); cmd != "" {
			out["install_command"] = cmd
		}
	}
	return out, nil
}

// installCommand — готовая команда установки из внутреннего репозитория.
// Пустая строка, если менеджер неизвестен: это не повод ронять карточку.
func (h *PackagesHandler) installCommand(row *repo.VersionRow) string {
	plugin, err := h.Registry.Get(row.Manager)
	if err != nil {
		return ""
	}
	ref := registry.Ref{
		Manager: row.Manager, Name: row.Name, DisplayName: row.DisplayName,
		Version: row.Version.Version, RawVersion: row.Version.RawVersion,
	}
	installBaseURL, installRepo := h.Cfg.ArtifactInstallLocation(row.Manager)
	return plugin.InstallCommand(ref, installBaseURL, installRepo)
}

// vulnerabilityViews — уязвимости для выдачи. full=true — карточка пакета, у
// неё полей больше; false — карточка заявки (порт request_payload, там
// affected_ranges и index_version_id не отдаются).
func vulnerabilityViews(vulns []domain.Vulnerability, full bool) []map[string]any {
	out := make([]map[string]any, 0, len(vulns))
	for _, v := range vulns {
		item := map[string]any{
			"id":             v.ExternalID,
			"score":          v.Score,
			"cvss_vector":    v.CVSSVector,
			"severity":       v.Severity,
			"url":            v.URL,
			"summary":        v.Summary,
			"fixed_versions": listOrEmpty(v.FixedVersions),
		}
		if full {
			item["affected_ranges"] = v.AffectedRanges
			item["index_version_id"] = v.VulnIndexVersionID
		}
		out = append(out, item)
	}
	return out
}

func artifactViews(artifacts []domain.Artifact) []map[string]any {
	out := make([]map[string]any, 0, len(artifacts))
	for _, a := range artifacts {
		out = append(out, map[string]any{
			"filename":      a.Filename,
			"nexus_url":     a.NexusURL,
			"sha256":        a.SHA256,
			"size_bytes":    a.SizeBytes,
			"s3_key":        a.StagingPath,
			"s3_deleted_at": a.StagingClearedAt,
			"status":        a.Status,
		})
	}
	return out
}

func codeFindingViews(findings []domain.CodeFinding) []map[string]any {
	out := make([]map[string]any, 0, len(findings))
	for _, f := range findings {
		out = append(out, map[string]any{
			"scanner":  f.Scanner,
			"rule_id":  f.RuleID,
			"severity": f.Severity,
			"message":  f.Message,
			"file":     f.FilePath,
			"line":     f.Line,
			"matched":  f.Matched,
		})
	}
	return out
}

// listOrEmpty — пустой список вместо null: фронтенд делает .map() по этим
// полям без проверки.
func listOrEmpty(v []string) []string {
	if v == nil {
		return []string{}
	}
	return v
}

func intParam(raw string, fallback int) (int, error) {
	if strings.TrimSpace(raw) == "" {
		return fallback, nil
	}
	return strconv.Atoi(raw)
}
